ta/arch/arm/link.mk

```diff
@@ -108,9 +116,15 @@ $(link-out-dir$(sm))/$(user-ta-uuid).ta: \
 			$(TA_SIGN_KEY) \
 			$(lastword $(SIGN_ENC))
 	@$(cmd-echo-silent) '  $$(cmd-echo$(user-ta-uuid)) $$@'
+ifeq ($(CFG_STITCH_TA),y)
+	$(q)$(SIGN_ENC) stitch --key $(TA_PUBLIC_KEY) $$(crypt-args$(user-ta-uuid)) \
+        --uuid $(user-ta-uuid) --sig $(TA_SIG_DIR)/$(user-ta-uuid).sig \
+		--algo $(TA_SIGN_ALG) --in $$< --out $$@
```

Comment:
Here the TA could be re-built but stitched with the signature from a likely previous _signatures/_ files build. You would need to recompute the signature once `$(user-ta-uuid).stripped.elf` is updated?

If you're considering using a key behind a pkcs11 token interface, `sing_encryp.py` could be updated to talk straight to the pkcs11 token for signing the TA(s). This would maybe remove the need for signature stitching.

Reply:
> Here the TA could be re-built but stitched with the signature from a likely previous signatures/ files build. You would need to recompute the signature once $(user-ta-uuid).stripped.elf is updated?

Yes, if an OP-TEE TA changes, we have to recompute the signature. However, I don't think this is very important, since the OP-TEE TAs remain the same within a release. If I update the OP-TEE to another version, there is more work to do than just recomputing the signatures. Plus there is the same issue with user TAs, which change as much as often.

> If you're considering using a key behind a pkcs11 token interface, sing_encryp.py could be updated to talk straight to the pkcs11 token for signing the TA(s). This would maybe remove the need for signature stitching.

Good idea, I see minor issues with it though. 
* In our case I have no access to the HSM (its on a remote serve), so I wouldn't be able to build the OP-TEE this way, and the guy who does have access to it, has very little time to deal with it. 
* Do you think we could implement this in a generic way so everyone could sign the TAs without modifying the `sign_encrypt.py` script? A possible solution would be that the user should provide a module which would implement some specific stuff related to the pkcs11 token interface.  
* If we were to go this way, I'd have to see when I could implement it, since I have burned through the hours I had for this topic.

Reply:
> > Here the TA could be re-built but stitched with the signature from a likely previous signatures/ files build. You would need to recompute the signature once $(user-ta-uuid).stripped.elf is updated?
> 
> Yes, if an OP-TEE TA changes, we have to recompute the signature. However, I don't think this is very important, since the OP-TEE TAs remain the same within a release. If I update the OP-TEE to another version, there is more work to do than just recomputing the signatures. Plus there is the same issue with user TAs, which change as much as often.

What do you mean with "changes"? If you get a different binary when recompiling the TA? If you expect the same binary each time, why bother recompiling it?

Reply:
> What do you mean with "changes"? If you get a different binary when recompiling the TA? If you expect the same binary each time, why bother recompiling it?

With "changes" I mean if the source code of a TA changes (a bug is fixed or new features are added) - happens only when a new version is released. In that case if we were to use an old signature, the stitching would fail. So we need to build the OP-TEE once, compute the hashes of the `<ta-uuid>.stripped.elf` files and sign them. Once done, the build process will always succeed, since the `<ta-uuid>.stripped.elf` files (and hence their hashes and signatures) do not change.

Reply:
It seems to me that you need to install pre-built/pre-signed TA binaries, rather than stitching some bin's for them to be installed.

Reply:
> It seems to me that you need to install pre-built/pre-signed TA binaries, rather than stitching some bin's for them to be installed.

What do you mean to install pre-built/pre-signed TA binaries? The build process does not allow that, as it always builds and tries to sign the TAs. Or you mean to let the build fail on purpose and to sign the `.stripped.elf` files afterwards?

The reason for the PR is the following: the offline-signing procedure requires that the key `./keys/default_ta.pem` is a public key. If that is the case, the OPTEE build process fails with error: `ERROR:sign_encrypt.py:Provided key cannot be used for signing, please use offline-signing mode.`. The reason for that is that the build process tries to sign the TAs with the public key and fails. 

To sum up: OPTEE OS cannot be compiled if `./keys/default_ta.pem` is a public key, what is required for offline-signing.

Reply:
I agree that OP-TEE should be able to build without mandating signature at build time.
Building with no `TA_SIGN_KEY` as far as at least **TA_PUBLIC_KEY** is defined (to get the public key embedded in Core) should not fail.